pennylane/templates/state_preparations.py

```diff
@@ -0,0 +1,315 @@
+# Copyright 2018 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+State Preparations
+==================
+
+**Module name:** :mod:`pennylane.templates.state_preparations`
+
+.. currentmodule:: pennylane.templates.state_preparations
+
+This module provides routines that prepare a given state using only
+elementary gates.
+
+Qubit architectures
+-------------------
+
+.. autosummary::
+
+    BasisStatePreparation
+    MottonenStatePreparation
+
+Code details
+^^^^^^^^^^^^
+"""
+import math
+from collections.abc import Iterable
+from scipy import sparse
+
+import numpy as np
+import pennylane as qml
+
+# pylint: disable=len-as-condition,arguments-out-of-order
+
+
+def gray_code(rank):
+    """Generates the Gray code of given rank.
+
+    Args:
+        rank (int): rank of the Gray code (i.e. number of bits)
+    """
+    return ["{0:0{1}b}".format(i ^ (i >> 1), rank) for i in range(0, 1 << rank)]
```

Comment:
Turns out, that this can be fastened by using a resursive approach (the one described here:
https://stackoverflow.com/questions/38738835/generating-gray-codes)

It really performs considerably better for larger systems (on my machine starting from around 20 qubits).

Simple script that can be nice to see how they perform:
```python
import time

def so_gray_code(n):
    def gray_code_recurse (g,n):
        k=len(g)
        if n<=0:
            return

        else:
            for i in range (k-1,-1,-1):
                char='1'+g[i]
                g.append(char)
            for i in range (k-1,-1,-1):
                g[i]='0'+g[i]

            gray_code_recurse (g,n-1)

    g=['0','1']
    gray_code_recurse(g,n-1)
    return g

def gray_code(rank):
    """Generates the Gray code of given rank.

    Args:
        rank (int): rank of the Gray code (i.e. number of bits)
    """
    return ["{0:0{1}b}".format(i ^ (i >> 1), rank) for i in range(0, 1 << rank)]

def func(n):
    start = time.time()
    a = gray_code(n)
    print(time.time()-start)
    b = so_gray_code(n)
    print(time.time()-start)
```
On my machine for ``rank==24`` ``gray_code`` ran for 7.7 seconds, ``so_gray_code`` for 3 seconds, at ``rank==20`` the difference is 0.2 seconds (pretty negligable).

Reply:
Thanks a lot for this!

I also thought that it might be advantageous to use lazy evaluation here, because the gray code has an exponential number of elements. For many qubits this will likely create a huge memory consumption to create the whole gray code at once.

Reply:
But I think we postpone this until people run out of memory. You will always have large memory consumption because you also use exponentially many gates, so I think the Gray code is not the most memory-demanding part of this anyways.